independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,117 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicReference<JarFileReference> jarFileReference;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
+
+    private volatile boolean releasing = false;
+
+    private JarFileReference(AtomicReference<JarFileReference> jarFileReference, JarFile jarFile) {
+        this.jarFileReference = jarFileReference;
+        this.jarFile = jarFile;
+    }
+
+    /**
+     * Increase the readers counter of the jarFile and returns it.
+     *
+     * @return The jarFile wrapped by this reference or null if a release has been requested thus making the file no longer
+     *         accessible.
+     */
+    private JarFile readLockAcquire() {
+        readersCounter.incrementAndGet();
+        if (releasing) {
```

Comment:
This is the race which can break things I mentioned at https://github.com/quarkusio/quarkus/pull/40942/files#r1626278120: 
what happen if we read the `release` when is not yet set by a close?